fabric-data-attachment-api-v1/src/main/java/net/fabricmc/fabric/impl/attachment/sync/AttachmentChange.java

```diff
@@ -0,0 +1,119 @@
+import net.minecraft.server.network.ServerPlayerEntity;
+import net.minecraft.util.Identifier;
+import net.minecraft.world.World;
+
+import net.fabricmc.fabric.api.attachment.v1.AttachmentType;
+import net.fabricmc.fabric.api.networking.v1.PacketByteBufs;
+import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking;
+import net.fabricmc.fabric.impl.attachment.AttachmentRegistryImpl;
+import net.fabricmc.fabric.impl.attachment.AttachmentTypeImpl;
+import net.fabricmc.fabric.impl.attachment.sync.s2c.AttachmentSyncPayload;
+import net.fabricmc.fabric.mixin.networking.accessor.ServerCommonNetworkHandlerAccessor;
+
+// empty optional for removal
+public record AttachmentChange(AttachmentTargetInfo<?> targetInfo, AttachmentType<?> type, byte[] data) {
+	public static final PacketCodec<PacketByteBuf, AttachmentChange> PACKET_CODEC = PacketCodec.tuple(
+			AttachmentTargetInfo.PACKET_CODEC, AttachmentChange::targetInfo,
+			Identifier.PACKET_CODEC.xmap(
+					id -> Objects.requireNonNull(AttachmentRegistryImpl.get(id)),
+					AttachmentType::identifier
+			), AttachmentChange::type,
+			PacketCodecs.BYTE_ARRAY, AttachmentChange::data,
+			AttachmentChange::new
+	);
+	private static final int MAX_PADDING_SIZE_IN_BYTES = AttachmentTargetInfo.MAX_SIZE_IN_BYTES + AttachmentSync.MAX_IDENTIFIER_SIZE_IN_BYTES;
+	// add a parameter?
+	private static final int MAX_DATA_SIZE_IN_BYTES = 0x10000 - MAX_PADDING_SIZE_IN_BYTES;
+
+	@SuppressWarnings("unchecked")
+	public static AttachmentChange create(AttachmentTargetInfo<?> targetInfo, AttachmentType<?> type, @Nullable Object value) {
+		PacketCodec<PacketByteBuf, Object> codec = (PacketCodec<PacketByteBuf, Object>) ((AttachmentTypeImpl<?>) type).packetCodec();
+		assert codec != null;
+
+		PacketByteBuf buf = PacketByteBufs.create();
+		buf.writeOptional(Optional.ofNullable(value), codec);
+		byte[] encoded = buf.array();
+
+		if (encoded.length >= MAX_DATA_SIZE_IN_BYTES) {
+			throw new IllegalArgumentException("Data for attachment '%s' was too big (%d bytes, over maximum %d)".formatted(
+					type.identifier(),
+					encoded.length,
+					MAX_DATA_SIZE_IN_BYTES
+			));
+		}
+
+		return new AttachmentChange(targetInfo, type, encoded);
+	}
+
+	public static void partitionForPackets(List<AttachmentChange> changes, ServerPlayerEntity player) {
+		Set<Identifier> supported = ((SupportedAttachmentsClientConnection) ((ServerCommonNetworkHandlerAccessor) player.networkHandler).getConnection())
+				.fabric_getSupportedAttachments();
+		List<AttachmentChange> packetChanges = new ArrayList<>();
+		int byteSize = Integer.BYTES + 1; // VarInts.MAX_BYTES
+
+		for (AttachmentChange change : changes) {
+			if (!supported.contains(change.type.identifier())) {
+				continue;
+			}
+
+			int size = MAX_PADDING_SIZE_IN_BYTES + change.data.length;
+
+			if (byteSize + size >= MAX_DATA_SIZE_IN_BYTES) {
+				ServerPlayNetworking.send(player, new AttachmentSyncPayload(packetChanges));
+				packetChanges = new ArrayList<>();
```

Comment:
why allocate a new List instead of clearing the existing one?

Reply:
I think I was a bit hazy on how the networking code works and afraid the object could still be allocated by then.